RadClass/analysis.py

```diff
@@ -0,0 +1,205 @@
+
+        Return: list of rows for data indexing
+        '''
+
+        # collect start index from tracked timestamp
+        start_i, = np.where(self.processor.timestamps == self.working_time)
+
+        # NOTE: this behavior is currently disabled by the if-statement in
+        # march() removing the 'or' portion with allow this to work
+        #
+        # if the final portion of the file is smaller than a full integration
+        # interval, only what is left is collected for this analysis
+        if start_i + self.integration >= len(self.processor.timestamps):
+            end_i = len(self.processor.timestamps) - 1
+        else:
+            end_i = start_i + self.integration
+
+        # enumerate number of rows to integrate exclusive of the endpoint
+        rows = np.arange(start_i, end_i)
+        return rows
+
+    def march(self):
+        '''
+        Advance a step using stride time to find the new working integration
+        interval. Also keeps track of the EOF state but checking for
+        out-of-index situations.
+
+        Returns: boolean for EOF
+        '''
+
+        # find the working integration interval starting index and advance stride
+        start_i, = np.where(self.processor.timestamps == self.working_time)
+        start_i = start_i[0]
+        new_i = start_i + self.stride
+
+        # stop analysis if EOF reached
+        # NOTE: stops prematurely, only does analysis for windows of full integration time
+        if new_i >= len(self.processor.timestamps) or (new_i + self.integration) >= len(self.processor.timestamps):
+            return False
+        else:
+            # update working integration interval timestep
+            self.working_time = self.processor.timestamps[new_i]
+            return True
+
+    def iterate(self):
+        '''
+        Full iteration over the entirety of the data file. Runs
+        until EOF reached. Prints progress over the course of the analysis. Only
+        runs for a set node (datapath) with data already queued into class.
+        '''
+
+        while self.running:
+            # print status at set intervals
+            if np.where(self.processor.timestamps == self.working_time)[0][0] % 100000 == 0:
+                readable_time = time.strftime('%m/%d/%Y %H:%M:%S',  time.gmtime(self.working_time))
+                print("=========================================================")
+                print("Currently working on timestamps: {}\n".format(readable_time))
+            
+            # execute analysis and advance in stride
+            rows = self.collect_rows()
+            data = self.collapse_data(rows)
+
+            # pass data to analysis object if available
+            if self.analysis is not None:
+                self.analysis.run(data)
+
+            if self.store_data:
+                self.storage = pd.concat([self.storage,pd.DataFrame([data], index = [self.working_time])])
```

Comment:
I may have to test this. There is conflicting discussions about whether `concat` is [faster](https://stackoverflow.com/questions/31690076/creating-large-pandas-dataframes-preallocation-vs-append-vs-concat) or [slower](https://stackoverflow.com/questions/36489576/why-does-concatenation-of-dataframes-get-exponentially-slower).

I certainly see why it would be a bottleneck, but I'm almost wondering if appending to a list or array (and then constructing a DataFrame from the list, or just saving the array to csv outright) would be faster than pre-allocating a DataFrame to fill.